HDDS-8490. [Snapshot] Ability to cancel an in-progress snapdiff job

xBis7 opened the pull request:
## What changes were proposed in this pull request?

This PR adds an option to the `ozone sh snapshot snapshotDiff` command, to cancel an IN_PROGRESS snapshotDiff job. If the option is used and the job is IN_PROGRESS, then the status is updated to CANCELED. 

The part of the code that might take up the most resources and cause the delay, has been refactored so that we can keep checking if the JobStatus is CANCELED before every method call. 

If the job is canceled, then the method doing the calculations returns and the job remains CANCELED until the `SnapshotDiffCleanupService` deletes it from the snapDiffJobTable and the user can resubmit it. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8490

## How was this patch tested?

This patch was tested with new unit and integration tests. It was also tested manually using the docker dev environment.

xBis7 commented:
@hemantk-12 @smengcl Can you take a look at this PR?

hemantk-12 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java:
```diff
@@ -0,0 +1,302 @@
+  @BeforeAll
+  public static void init() throws AuthenticationException,
+      IOException, RocksDBException {
+    metaDir = GenericTestUtils.getRandomizedTestDir();
```
You can use @TempDir annotation to create temp dir for the test.

xBis7 replied:
@hemantk-12 I had multiple workflows failing due to `@tempdir` not being able to cleanup properly and I found online that other people had the same experience. 

These are some of the failing workflows. They all belong to a dummy branch I had for cancelling a snapshot diff job.

https://github.com/xBis7/ozone/actions/runs/5147594598/jobs/9268500272#step:6:2565

https://github.com/xBis7/ozone/actions/runs/5147452987/jobs/9267857940#step:6:2675

https://github.com/xBis7/ozone/actions/runs/5143362416/jobs/9258280943#step:6:2562

hemantk-12 replied:
I'm surprised and curious what is causing this failure. `@TempDir` is used at some many other place. Were you trying to cleanup or delete it?

xBis7 replied:
> Were you trying to cleanup or delete it?

No, the annotation is supposed to take care of creation and cleanup. Check below,

https://github.com/xBis7/ozone/blob/a575a094515f75efe95b9dfda8c9a8bf1de0eb3a/hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java#L69-L70

https://github.com/xBis7/ozone/blob/0694343dae2a3f0237eae4e30b0e5359dde3da75/hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java#L75-L76

hemantk-12 reviewed (commented):
Thanks for the patch @xBis7.

Overall looks good to me. Left some comments.

xBis7 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:
```diff
@@ -627,6 +628,47 @@ public void testSnapDiff() throws Exception {
+    String toSnapName = "snap-2-" + RandomStringUtils.randomNumeric(5);
+    createSnapshot(volumeName, bucketName, toSnapName);
+
+    // Cancel works only if the job is IN_PROGRESS, and
+    // it's ignored if the job has any other status.
```
I added it to explain it doesn't make any difference to remove cancel a new job. I removed it.

xBis7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -1010,6 +1066,11 @@ private long generateDiffReport(
       try (ClosableIterator<byte[]>
                objectIdsIterator = objectIDsToCheck.iterator()) {
         while (objectIdsIterator.hasNext()) {
+          if (snapDiffJobTable.get(jobKey).getStatus()
```
> To avoid NPE.

Yes, the job might have failed or being rejected and got removed from the table.

> Should we check after like every 100 entries?

Sure, I'll fix that.

xBis7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -718,36 +746,63 @@ private void generateSnapshotDiffReport(final String jobKey,
+          },
+          () -> {
+            if (bucketLayout.isFileSystemOptimized()) {
+              validateSnapshotsAreActive(volumeName, bucketName,
```
Done.

xBis7 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:
```diff
@@ -627,6 +628,47 @@ public void testSnapDiff() throws Exception {
+    // Executing the command again should return CANCELED,
+    // until the job is picked up by the SnapshotDiffCleanupService
+    // and removed from the snapDiffJobTable.
+    response = store.snapshotDiff(volumeName,
```
@hemantk-12 This is testing snapDiff cancel from a client perspective using the ObjectStore. IMO, this check is more suitable for `TestSnapshotDiffManager` which is checking tables and such. I placed it there. Let me know what you think.

hemantk-12 replied:
I agree and disagree at the same time :P. The reason I wanted it here because this is integration test and should test end-end flow. While `TestSnapshotDiffManager` is unit test and should just test `SnapshotDiffManager` as unit.

xBis7 replied:
I can add it here as well. IMO, they are both integration tests since in `TestSnapshotDiffManager` we are not mocking anything, we are using `TestOmManagers` which creates an actual OzoneManager. The difference is that this test spins up a whole cluster to run the test on.

We had a previous discussion in a sync call about this. Most of Ozone tests are integration tests, but we keep calling them unit tests because they don't use MiniOzoneCluster.

xBis7 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java:
```diff
@@ -0,0 +1,322 @@
+    // Job is still in the snapDiffJobTable,
+    // so it shouldn't exist in the purged job table.
+    Assertions.assertNull(snapshotDiffCleanupService
+        .getEntryFromPurgedJobTable(jobId));
```
@hemantk-12 I added the purged job table check here.

xBis7 commented:
@hemantk-12 Thanks for reviewing this, I've addressed all your comments.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -422,6 +442,12 @@ public SnapshotDiffResponse getSnapshotDiffReport(
+          new SnapshotDiffReportOzone(snapshotRoot.toString(), volumeName,
+              bucketName, fromSnapshotName, toSnapshotName, new ArrayList<>(),
+              null),
+          CANCELED, defaultWaitTime);
```
Return zero instead of `defaultWaitTime` here.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -389,9 +399,19 @@ public SnapshotDiffResponse getSnapshotDiffReport(
 
+    // Check if the job already exists in the table
+    // before getting report status.
+    if (cancel && Objects.nonNull(snapDiffJobTable.get(snapDiffJobKey)) &&
```
There are few problems by adding this to current creation flow.
1. If job doesn't exist, we would create a new job. Which is not needed. Correct behavior should be return empty response with a message.
1. If job is not `IN_PROGRESS` status and it is `DONE`, it will not mark the job cancel and will return first page response. I am not sure if that is correct response to client.
1. It will fail at line 409, if two or more commands are issued at the same time for the same snapDiff job because one of them will mark it cancel successfully and other/s will fail due to [check](https://github.com/apache/ozone/blob/9034434285f3877246b3ebc6a7513c082931b279/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java#L1164).

I think it would be better to create a separate flow for cancel and handle these cases appropriately. Also snapDiff job submission and cancellation should be consider two separate options in `OmSnapshotManager` and `SnapshotDiffManager`.

In [OzoneManager](https://github.com/apache/ozone/pull/4819/files#diff-54bbfc5b5dd89a4d035bee206c53d3d11bbb579a2957abca589e313283d98af9R4523), you can use cancel to decide which flow needs to be called.

```
  public SnapshotDiffResponse snapshotDiff(String volume,
                                           String bucket,
                                           String fromSnapshot,
                                           String toSnapshot,
                                           String token,
                                           int pageSize,
                                           boolean forceFullDiff,
                                           boolean cancel)
      throws IOException {
    if (cancel) {
      return omSnapshotManager.cancelSnapshotDiff(volume, bucket,
          fromSnapshot, toSnapshot);
    } else {
      return omSnapshotManager.getSnapshotDiffReport(volume, bucket,
          fromSnapshot, toSnapshot, token, pageSize, forceFullDiff);
    }
  }
```

xBis7 replied:
> Also snapDiff job submission and cancellation should be consider two separate options in OmSnapshotManager and SnapshotDiffManager.
> 
> In [OzoneManager](https://github.com/apache/ozone/pull/4819/files#diff-54bbfc5b5dd89a4d035bee206c53d3d11bbb579a2957abca589e313283d98af9R4523), you can use cancel to decide which flow needs to be called.

The OzoneManager changes will certainly make everything look cleaner but won't change anything.

> 1. If job doesn't exist, we would create a new job. Which is not needed. Correct behavior should be return empty response with a message.

Are you saying that if the job doesn't exist then cancel should just return and not proceed with submitting the job? Maybe return CANCELED response? Why submit a new job with a cancel parameter? Or should we check for the case that the job is stuck as QUEUED?

> 2. If job is not IN_PROGRESS status and it is DONE, it will not mark the job cancel and will return first page response. I am not sure if that is correct response to client.

I thought that's the intention here. The user wants to cancel a job only if it's in progress. If the job is DONE, then why cancel it? The system has already gone through the whole process and finished with it. 

> 3. It will fail at line 409, if two or more commands are issued at the same time for the same snapDiff job because one of them will mark it cancel successfully and other/s will fail due to [check](https://github.com/apache/ozone/blob/9034434285f3877246b3ebc6a7513c082931b279/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java#L1164).

If the job is not IN_PROGRESS then it shouldn't be marked as canceled. I don't think the exception affects us in any case. 

We try to update the job status and
* <b>cancel succeeds</b>, job status is updated to CANCELED. 
Cancel check should pick it up but in any case where somehow we end up updating status to DONE or FAILED or REJECTED, we get an exception, status is not updated and the job is left in the table as CANCELED. In that case The cleanup service picks it up and deletes everything.
* <b>cancel fails</b>, job status is not updated to CANCELED. 
Job status is either FAILED, REJECTED or DONE. If it's DONE then great it will return the diff report to the user. If it's one of the other options, the job will be cleaned up by the cleanup service.

The way I see the flow, we run `getSnapshotDiffReport()` with cancel=true. 
If the job is
* NEW, cancel is ignored, job is registered as QUEUED and then submitted
* QUEUED, cancel is ignored, gets submitted again
* IN_PROGRESS, cancel takes effect, job is marked as canceled
* DONE, cancel is ignored, get the diff report 
* FAILED - REJECTED - CANCELED, cancel is ignored, get the response until the cleanup service deletes the job.

This all seems intended behavior to me.

hemantk-12 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -718,36 +746,60 @@ private void generateSnapshotDiffReport(final String jobKey,
-          objectIdToKeyNameMapForToSnapshot);
+      // Check if the job is cancelled, before every method call.
+      for (Callable<Void> methodCall : methodCalls) {
+        if (snapDiffJobTable.get(jobKey).getStatus()
+            .equals(CANCELED)) {
+          return;
+        }
```
Can you please move this to another function `validateSnapshotDiffJobIsActive`? An reuse it at other place like lines 1071-1073 and lines 784-786.

xBis7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffJob.java:
```diff
@@ -61,6 +62,7 @@ public SnapshotDiffJob(long creationTime,
                          String fromSnapshot,
                          String toSnapshot,
                          boolean forceFullDiff,
+                         boolean cancel,
```
There was an idea about using the information when loading jobs on startup. But now this information seems redundant. I'll clean this up.

xBis7 commented:
@hemantk-12 I've addressed all your comments and updated the patch to send different responses back to the client in case job cancelling fails. I've also updated the existing tests and added new ones as well. Let me know how it looks.

smengcl commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/ObjectStore.java:
```diff
@@ -670,15 +670,17 @@ private List<OzoneSnapshot> getNextListOfSnapshots(String prevSnapshot)
    * @return the difference report between two snapshots
```
new param `cancel` should be added to the javadoc as well.

smengcl commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java:
```diff
@@ -1070,10 +1070,11 @@ List<OzoneSnapshot> listSnapshot(
    * @return the difference report between two snapshots
```
and here

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OzoneManagerProtocol.java:
```diff
@@ -717,13 +717,15 @@ default List<SnapshotInfo> listSnapshot(
    * @return the difference report between two snapshots
```
Same here

smengcl commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/SnapshotDiffHandler.java:
```diff
@@ -60,6 +60,13 @@ public class SnapshotDiffHandler extends Handler {
 
+
+  @CommandLine.Option(names = {"-c", "--cancel"},
+      description = "Cancel a running snapshotDiff job. " +
+          "If the job is not IN_PROGRESS then the option will fail.",
```
```suggestion
      description = "Request to cancel a running SnapshotDiff job. " +
          "If the job is not IN_PROGRESS, the request will fail.",
```

smengcl commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -1799,10 +1800,22 @@ message SnapshotDiffResponse {
+    JOB_DONE = 3;
+    INVALID_STATUS_TRANSITION = 4;
+    JOB_ALREADY_CANCELED = 5;
+    CANCEL_SUCCESS = 6;
```
nit
```suggestion
    CANCELLATION_SUCCESS = 6;
```

smengcl commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -1799,10 +1800,22 @@ message SnapshotDiffResponse {
   }
+
+  enum CancelStatusProto {
+    JOB_NOT_CANCELED = 1;
```
Let's use double-l
```suggestion
    JOB_NOT_CANCELLED = 1;
```

smengcl commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -1799,10 +1800,22 @@ message SnapshotDiffResponse {
+    CANCELED = 6;
   }
+
+  enum CancelStatusProto {
```
IMO this is more of a `Result` to the one-time client request rather than a `Status` for the diff job.

```suggestion
  enum JobCancelResultProto {
```

e.g. https://github.com/apache/ozone/blob/bdd26eeca9fe7adfd87a4ee1e0acfd84d6148c57/hadoop-hdds/interface-server/src/main/proto/ScmServerProtocol.proto#L182-L199

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/snapshot/SnapshotDiffResponse.java:
```diff
@@ -43,16 +45,59 @@ public static JobStatus fromProtobuf(JobStatusProto jobStatusProto) {
+  /**
+   * Snapshot diff cancel status enum.
+   */
+  public enum CancelStatus {
```
Similarly
```suggestion
  public enum JobCancelResult {
```

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/snapshot/SnapshotDiffResponse.java:
```diff
@@ -67,18 +112,28 @@ public long getWaitTimeInMs() {
+      } else {
+        str.append("Snapshot diff job is ");
+        str.append(jobStatus);
+        str.append("\n");
```
nit
```suggestion
        str.append(". ");
```

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/snapshot/SnapshotDiffResponse.java:
```diff
@@ -43,16 +45,59 @@ public static JobStatus fromProtobuf(JobStatusProto jobStatusProto) {
+  public enum CancelStatus {
+    JOB_NOT_CANCELED("Job is not canceled"),
+    NEW_JOB("Cannot cancel a newly submitted job"),
+    JOB_DONE("Job is DONE, cancel failed"),
```
```suggestion
    JOB_DONE("Job is already DONE"),
```

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/snapshot/SnapshotDiffResponse.java:
```diff
@@ -43,16 +45,59 @@ public static JobStatus fromProtobuf(JobStatusProto jobStatusProto) {
+   * Snapshot diff cancel status enum.
+   */
+  public enum CancelStatus {
+    JOB_NOT_CANCELED("Job is not canceled"),
```
IIUC this indicates "Job has never been cancelled"?

xBis7 replied:
Yes, I'll change the description to `Job hasn't been cancelled` to make it more explanatory.

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDiffCleanupService.java:
```diff
@@ -124,6 +125,16 @@ public void run() {
+      return db.get().get(snapDiffPurgedJobCfh,
+          codecRegistry.asRawData(jobId));
+    } catch (IOException | RocksDBException e) {
+      throw new RuntimeException(e);
```
```suggestion
      // TODO: [SNAPSHOT] Fail gracefully.
      throw new RuntimeException(e);
```

smengcl commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:
```diff
@@ -631,6 +639,127 @@ public void testSnapDiff() throws Exception {
+    String toSnapName = "snap-2-" + RandomStringUtils.randomNumeric(5);
+    createSnapshot(volumeName, bucketName, toSnapName);
+
+    SnapshotDiffResponse response = store.snapshotDiff(
+        volumeName, bucketName, fromSnapName, toSnapName,
+        null, 0, false, false);
+
+    assertEquals(IN_PROGRESS, response.getJobStatus());
+
+    response = store.snapshotDiff(volumeName,
+        bucketName, fromSnapName, toSnapName,
+        null, 0, false, true);
+
+    // Job status should be updated to CANCELED.
+    assertEquals(CANCELED, response.getJobStatus());
```
Could this be a source of flakiness? There is no guarantee that the diff job is still running when calling the second `snapshotDiff` with `cancel = true` right?

xBis7 replied:
> Can fix this if there is a way to suspend the SnapshotDiff worker thread before the first call.

That's a good idea, but the first call is submitting the job and we would be suspending the thread for a job that doesn't even exist in the snap diff table. If we do it after the submitting then there is no difference with cancelling the job. 

The only way I can think of is calling the methods that perform all these operations and do the suspension between the calls instead of using the ObjectStore api.

smengcl replied:
Hmm. I'm think if we could intercept the `snapDiffExecutor`:

https://github.com/apache/ozone/blob/9c6cd4b93e0427fc12bbedae8156c746cbe6cea1/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java#L216-L224

Or just add a spin lock just for the tests inside `generateSnapshotDiffReport()`.

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:
```diff
@@ -489,6 +489,15 @@ public static String getOzonePathKeyWithVolumeBucketNames(
+  @VisibleForTesting
+  public SnapshotDiffCleanupService getSnapshotDiffCleanupService() {
+    return snapshotDiffCleanupService;
+  }
```
```suggestion
  }

```

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -418,6 +478,12 @@ public SnapshotDiffResponse getSnapshotDiffReport(
               bucketName, fromSnapshotName, toSnapshotName, new ArrayList<>(),
               null),
           REJECTED, defaultWaitTime);
+    case CANCELED:
+      return new SnapshotDiffResponse(
+          new SnapshotDiffReportOzone(snapshotRoot.toString(), volumeName,
+              bucketName, fromSnapshotName, toSnapshotName, new ArrayList<>(),
+              null),
+          CANCELED, 0L, CancelStatus.CANCEL_SUCCESS);
```
Unrelated to this PR but I think `SnapshotDiffReportOzone` could use a `Builder` subclass. cc @hemantk-12

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -623,20 +689,29 @@ private synchronized SnapshotDiffJob getSnapDiffReportStatus(
-                                          final String bucketName,
-                                          final String fromSnapshotName,
-                                          final String toSnapshotName)
+  private boolean validateDiffJobAndSnapshotsAreActive(
```
```suggestion
  private boolean areDiffJobAndSnapshotsActive(
```

since it returns `boolean` now

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:
```diff
@@ -710,43 +787,67 @@ private void generateSnapshotDiffReport(final String jobKey,
-          objectIDsToCheckMap,
-          objectIdToKeyNameMapForFromSnapshot,
-          objectIdToKeyNameMapForToSnapshot);
+      // These are the most time and resource consuming method calls.
+      // Split the calls into steps and store them in an array, to avoid
+      // repetition while constantly checking if the job is cancelled.
+      Callable<Void>[] methodCalls = new Callable[]{
+          () -> {
```
nice

xBis7 replied:
thanks

xBis7 commented:
@smengcl Thanks for review, I've addressed all your comments and made the changes. Pending the comment about the test and its flakiness.

smengcl reviewed (approved):
lgtm pending CI

smengcl commented:
Thanks @xBis7 for the PR. Thanks @hemantk-12 for reviewing this.

We can file a follow-up jira to deal with the potential flakiness if needed.

xBis7 commented:
@smengcl @hemantk-12 Thanks for the reviews.
